### PR TITLE
More refactoring to obey platform abstraction lint

### PR DESCRIPTION
**Pull request by brson**
The most interesting things here are moving `std/sys/common` to `std/sys_common`, and `std/num/{f32,f64}.rs` to `std/{f32,f64}.rs`, and adding more documentation to `std/lib.rs`.

r? @alexcrichton

### CODE REVIEW

File: src/libstd/sys_common/mod.rs [hunk 1]
```diff
@@ -27,18 +27,6 @@
...
-    ($($t:tt)*) => (::sys_common::util::abort(format_args!($($t)*)))
-}
-
-macro_rules! rtassert {
-    ($e:expr) => ({
-        if !$e {
-            rtabort!(concat!("assertion failed: ", stringify!($e)))
-        }
-    })
-}
```
Comment (brson) [inline src/libstd/sys_common/mod.rs]: This macro is no longer used anywhere.

**Comment by brson** [comment]
A lot of this is just general cleanup...

**Comment by alexcrichton** [comment]
>  src/libstd/num/f64.rs → src/libstd/f64.rs 

How dare you!

File: src/libstd/lib.rs [hunk 2]
```diff
@@ -210,8 +210,27 @@
...
+// Tell the compiler to link to either panic_abort or panic_unwind
 #![needs_panic_runtime]
 
+// Always use alloc_system during stage0 since jemalloc might be unavailable or
+// disabled (Issue #30592)
+#![cfg_attr(stage0, feature(alloc_system))]
+
+// Turn warnings into errors, but only after stage0, where it can be useful for
+// code to emit warnings during language transitions
+#![cfg_attr(not(stage0), deny(warnings))]
```
Comment (alexcrichton) [inline src/libstd/lib.rs]: If you're *really* ambitious, I've been meaning to remove all the `cfg_attr` here and just leave `deny(warnings)`

File: src/libstd/lib.rs [hunk 3]
```diff
@@ -425,46 +403,62 @@ pub use core::u16;
 pub use core::u32;
 #[stable(feature = "rust1", since = "1.0.0")]
 pub use core::u64;
+#[stable(feature = "rust1", since = "1.0.0")]
+pub use alloc::boxed;
```
Comment (alexcrichton) [inline src/libstd/lib.rs]: I think the ordering here affects the ordering in rustdoc, right? (just confirming you want to change that too)

File: src/libstd/macros.rs [hunk 4]
```diff
@@ -184,6 +184,10 @@ macro_rules! assert_approx_eq {
     })
 }
 
+macro_rules! rtabort {
```
Comment (alexcrichton) [inline src/libstd/macros.rs]: aww I was hoping we could keep this contained within the `sys` module

File: src/libstd/sys_common/mod.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libstd/sys_common/mod.rs]: 🎺 ⚰ 🎺

File: src/libstd/thread/local.rs [hunk 5]
```diff
@@ -310,154 +310,8 @@ impl<T: 'static> LocalKey<T> {
...
-        if cfg!(target_os = "macos") {
-            ptr::read((*ptr).inner.get());
-        } else {
-            ptr::drop_in_place((*ptr).inner.get());
-        }
-    }
-}
-
 #[doc(hidden)]
+#[cfg(not(target_thread_local))]
```
Comment (alexcrichton) [inline src/libstd/thread/local.rs]: I think we'll want to always compile in this module. The theory behind this is that we compile the standard library with OS-based TLS (e.g. this module) but then *consumers* of the standard library can use the "fast TLS". The primary motivation here is OSX 10.6 and 10.7+, where 10.6 has slow TLS only but 10.7 has fast. We could in theory compile a 10.6 standard library but everyone else would have fast TLS.

**Comment by alexcrichton** [comment]
> moving std/sys/common to std/sys_common

I kinda liked having everything contained in one nice "sys" directory, but I guess not enough to go against conventions of filesystem hierarchies and module structures.

File: src/libstd/thread/local.rs (hunk 5 above)
Comment (brson) [inline src/libstd/thread/local.rs]: OK.

**Comment by brson** [comment]
Updated to leave os-specific TLS key in with documentation for why, move rtabort back to sys_common.

**Comment by alexcrichton** [comment]
@bors: retry

* ??

**Comment by alexcrichton** [comment]
ping @brson (for an update)

**Pull request merged**